src/constraints/shake.jl

```diff
@@ -1,141 +1,182 @@
+iters_var(sr::SHAKE_RATTLE) = value(sr.stats[2])
+iters_min(sr::SHAKE_RATTLE) = value(sr.stats[3].min)
+iters_max(sr::SHAKE_RATTLE) = value(sr.stats[3].max)
+iters_nmin(sr::SHAKE_RATTLE) = value(sr.stats[3].nmin)
+iters_nmax(sr::SHAKE_RATTLE) = value(sr.stats[3].nmax)
+  
+function setup_constraints!(sr::SHAKE_RATTLE, neighbor_finder, backend)
+
+    # Disable Neighbor interactions that are constrained
+    if typeof(neighbor_finder) != NoNeighborFinder
+        disable_constrained_interactions!(neighbor_finder, sr.clusters12)
+        disable_constrained_interactions!(neighbor_finder, sr.clusters23)
+        disable_constrained_interactions!(neighbor_finder, sr.clusters34)
+        disable_constrained_interactions!(neighbor_finder, sr.angle_clusters)
+    end
 
-                # Difference between unconstrainted velocities
-                v_k1k2 = sys.velocities[k2] .- sys.velocities[k1]
+    # Move to proper backend
+    # Assume only other backend is CPU, in which case we do nothing.
+    if typeof(backend) <: KernelAbstractions.GPU
 
-                err = abs(dot(r_k1k2, v_k1k2))
-                if err > ca.vel_tolerance
-                    # Re-arrange constraint equation to solve for Lagrange multiplier
-                    # This has a factor of dt which cancels out in the velocity update
-                    λₖ = -dot(r_k1k2, v_k1k2) / (dot(r_k1k2, r_k1k2) * (inv_m1 + inv_m2))
+        clusters12_gpu = []; clusters23_gpu = []
+        clusters34_gpu = []; angle_clusters_gpu = []
 
-                    # Correct velocities
-                    sys.velocities[k1] -= inv_m1 .* λₖ .* r_k1k2
-                    sys.velocities[k2] += inv_m2 .* λₖ .* r_k1k2
-                end
-            end
+        if length(sr.clusters12) > 0
+            clusters12_gpu = allocate(backend, eltype(sr.clusters12), length(sr.clusters12))
+            copy!(clusters12_gpu, sr.clusters12)
+        end
+        if length(sr.clusters23) > 0
+            clusters23_gpu = allocate(backend, eltype(sr.clusters23), length(sr.clusters23))
+            copy!(clusters23_gpu, sr.clusters23)
+        end
+        if length(sr.clusters34) > 0
+            clusters34_gpu = allocate(backend, eltype(sr.clusters34), length(sr.clusters34))
+            copy!(clusters34_gpu, sr.clusters34)
+        end
+        if length(sr.angle_clusters) > 0
+            angle_clusters_gpu = allocate(backend, eltype(sr.angle_clusters), length(sr.angle_clusters))
+            copy!(angle_clusters_gpu, sr.angle_clusters)
         end
 
-        converged = check_velocity_constraints(sys, ca)
+        sr = SHAKE_RATTLE(sr, clusters12_gpu, clusters23_gpu, clusters34_gpu, angle_clusters_gpu)
     end
-    return sys
+
+    return sr, neighbor_finder
+
 end
 
+
 """
     check_position_constraints(sys, constraints)
 
 Checks if the position constraints are satisfied by the current coordinates of `sys`.
 """
 function check_position_constraints(sys, ca::SHAKE_RATTLE)
```

Comment:
If these functions aren't specific to `SHAKE_RATTLE` then the argument type can be relaxed and they can be moved to constraints.jl.